pkg/gcpcredential/gcpcredential.go

```diff
@@ -0,0 +1,266 @@
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gcpcredential
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"os/exec"
+	"runtime"
+	"strings"
+	"time"
+
+	"k8s.io/cloud-provider-gcp/pkg/credentialconfig"
+	"k8s.io/klog/v2"
+)
+
+const (
+	metadataURL        = "http://metadata.google.internal./computeMetadata/v1/"
+	metadataAttributes = metadataURL + "instance/attributes/"
+	// DockerConfigKey is the URL of the dockercfg metadata key used by DockerConfigKeyProvider.
+	DockerConfigKey = metadataAttributes + "google-dockercfg"
+	// DockerConfigURLKey is the URL of the dockercfg metadata key used by DockerConfigURLKeyProvider.
+	DockerConfigURLKey = metadataAttributes + "google-dockercfg-url"
+	serviceAccounts    = metadataURL + "instance/service-accounts/"
+	metadataScopes     = metadataURL + "instance/service-accounts/default/scopes"
```

Comment:
Should consider composing these using smaller bits. Eg.
metadataScopes     = serviceAccounts + "default/scopes"
metadataToken       = serviceAccounts + "default/token"
allows you than change the service accounts path in 1 place, if you ever needed to change it. It also ensures that the paths up to that point are consistent.

Reply:
Ah, good catch - fixed